pkg/fetch/windows/product/product.go

```diff
@@ -0,0 +1,122 @@
+package product
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"path/filepath"
+
+	"github.com/cheggaaa/pb/v3"
+	"github.com/pkg/errors"
+
+	"github.com/MaineK00n/vuls-data-update/pkg/fetch/util"
+	utilhttp "github.com/MaineK00n/vuls-data-update/pkg/fetch/util/http"
+)
+
+const dataURL = "https://api.msrc.microsoft.com/sug/v2.0/sugodata/v2.0/en-US/affectedProduct?$skip=0"
+
+type options struct {
+	dataURL string
+	dir     string
+	retry   int
+}
+
+type Option interface {
+	apply(*options)
+}
+
+type dataURLOption string
+
+func (u dataURLOption) apply(opts *options) {
+	opts.dataURL = string(u)
+}
+
+func WithDataURL(url string) Option {
+	return dataURLOption(url)
+}
+
+type dirOption string
+
+func (d dirOption) apply(opts *options) {
+	opts.dir = string(d)
+}
+
+func WithDir(dir string) Option {
+	return dirOption(dir)
+}
+
+type retryOption int
+
+func (r retryOption) apply(opts *options) {
+	opts.retry = int(r)
+}
+
+func WithRetry(retry int) Option {
+	return retryOption(retry)
+}
+
+func Fetch(opts ...Option) error {
+	options := &options{
+		dataURL: dataURL,
+		dir:     filepath.Join(util.CacheDir(), "fetch", "windows", "product"),
+		retry:   3,
+	}
+
+	for _, o := range opts {
+		o.apply(options)
+	}
+
+	if err := util.RemoveAll(options.dir); err != nil {
+		return errors.Wrapf(err, "remove %s", options.dir)
+	}
+
+	log.Printf("[INFO] Fetch Microsoft Product")
+
+	bar := pb.StartNew(0)
+	u := options.dataURL
+	for u != "" {
+		r, err := func() (*response, error) {
+			resp, err := utilhttp.NewClient(utilhttp.WithClientRetryMax(options.retry)).Get(u)
+			if err != nil {
+				return nil, errors.Wrap(err, "fetch product")
+			}
+			defer resp.Body.Close() //nolint:errcheck
+
+			if resp.StatusCode != http.StatusOK {
+				_, _ = io.Copy(io.Discard, resp.Body)
+				return nil, errors.Errorf("error response with status code %d", resp.StatusCode)
+			}
+
+			var r response
+			if err := json.NewDecoder(resp.Body).Decode(&r); err != nil {
+				return nil, errors.Wrap(err, "decode json")
+			}
+
+			return &r, nil
+		}()
+		if err != nil {
+			return errors.Wrap(err, "fetch")
+		}
+
+		bar.SetTotal(int64(r.OdataCount))
+
+		for _, v := range r.Value {
+			splitted, err := util.Split(v.ID, "-", "-", "-", "-")
+			if err != nil {
+				return errors.Errorf("unexpected ID format. expected: %q, actual: %q", "00000000-0000-0000-0000-000000000000", v.ID)
+			}
+			if err := util.Write(filepath.Join(options.dir, splitted[0], splitted[1], splitted[2], splitted[3], splitted[4], fmt.Sprintf("%s.json", v.ID)), v); err != nil {
```

Comment:
With this directory structure, there is only one file per directory. Can we remove the last element `splitted[4]` from directory path?

Reply:
fixed: https://github.com/MaineK00n/vuls-data-update/compare/1fa0b06b57ebf7f7a1c08b6ac21614fba9dbcfec..df2f4ff7b43481434d6581253880e8fa5cf8591a